README.md
# hetzner dns ansible collection

Manage DNS records using ansible. E.g. [set DNS records while creating servers in the same play](https://git.osuv.de/m/hetzner_dns/wiki/Home).  
Or use hetzner DNS as an inventory for your playbooks.

* https://galaxy.ansible.com/markuman/hetzner_dns
* `ansible-galaxy collection install markuman.hetzner_dns`

## Auth

The API token can be set via ansible parameter or ENV variable.

| **Ansible Parameter** | **ENV Variable** |
| --- | --- |
| `api_token` (_alias `access_token`_) | `HETZNER_DNS_TOKEN` |

All modules supports and respects `check_mode: yes`.

## record

* `state: absent` 
    * _deleting_ a record: `name` and `type` are sufficient (_`ttl` doesn't matter_).
    * When `value` is not given, all records that match `name` and `type` will be removed!
    * When `value` is given, only one record will be deleted that matches `name`, `type`, and `value`.
* `state: present`
    * the `ttl` default value is set to `0` (_same as hetzner webinterface did._)
    * Either `zone_id` or `zone_name`  must be given, but not both.  

| parameters | default | comment |
| --- | --- | --- |
| `name` | - | name of a record |
| `purge` | `true` | Wether a existing value should be replaced (`true`) or appended (`false`) to achive multiple DNS records for one name. Alias parameter are: `replace`, `overwrite`, `solo`.
| `value` | - | required with `state: present` |
| `type` | - | type of record. valid records: "A" "AAAA" "NS" "MX" "CNAME" "RP" "TXT" "SOA" "HINFO" "SRV" "DANE" "TLSA" "DS" "CAA" |
| `ttl` | `300` | TTL of a record |
| `zone_id` | - | Required one of `zone_name` or `zone_id` |
| `zone_name` | - | Required one of `zone_name` or `zone_id` |
| `api_token` | - | Can be also set as env variable `HETZNER_DNS_TOKEN` |


```yml
- name: add record
  markuman.hetzner_dns.record:
    zone_name: osuv.de
    name: hetzner_dns_ansible_collection
    value: osuv.de.
    type: CNAME
    ttl: 300
  register: RECORD

- name: add record no change
  markuman.hetzner_dns.record:
    zone_name: osuv.de
    name: hetzner_dns_ansible_collection
    value: osuv.de.
    type: CNAME
    ttl: 300
  register: RECORD

- name: add record change
  markuman.hetzner_dns.record:
    zone_name: osuv.de
    name: hetzner_dns_ansible_collection
    value: osuv.de.
    type: CNAME
    ttl: 60
  register: RECORD

- name: del record
  markuman.hetzner_dns.record:
    zone_name: osuv.de
    name: hetzner_dns_ansible_collection
    type: CNAME
    state: absent
  register: RECORD

- name: add single A record
  markuman.hetzner_dns.record:
    zone_name: osuv.de
    name: dns
    value: 1.1.1.1
    type: A
    ttl: 60

- name: append A record
  markuman.hetzner_dns.record:
    zone_name: osuv.de
    name: dns
    value: 8.8.8.8
    type: A
    purge: no
    ttl: 60

- name: append A record
  markuman.hetzner_dns.record:
    zone_name: osuv.de
    name: dns
    value: 8.8.4.4
    type: A
    purge: no
    ttl: 60

- name: delete single A record
  markuman.hetzner_dns.record:
    zone_name: osuv.de
    name: dns
    value: 8.8.8.8
    type: A
    state: absent

- name: > 
    delete all A records
    when no `value` is given
  markuman.hetzner_dns.record:
    zone_name: osuv.de
    name: dns
    type: A
    state: absent

```

#### mx priority

When dialing with sevaral mx records, the priority must be part of the value.

```yml
- name: take care about mx record
  markuman.hetzner_dns.record:
    api_token: "{{ API }}"
    zone_name: osuv.de
    name: "@"
    value: 10 mail.osuv.de.
    type: MX
    ttl: 300
  register: RECORD
```

## record_info

When no filter is given, all records will be returned.  
Either `zone_id` or `zone_name`  must be given, but not both.  
Parameter `filter` must be a list, where each object supports the same keys as in `record`  module for search.

| parameter | default | comments |
| --- | --- | --- |
| `filter` | - | Apply a list of key/value pairs to search for |
| `zone_id` | - | Required one of `zone_name` or `zone_id` |
| `zone_name` | - | Required one of `zone_name` or `zone_id` |
| `api_token` | - | Can be also set as env variable `HETZNER_DNS_TOKEN` |


```yml
- name: fetch zone info
  markuman.hetzner_dns.record_info:
    filter:
      - name: fritzbox
        value: 192.168.178.1
        type: A
      - name: fritzbox
        type: AAAA
    zone_name: osuv.de
```

## zone_info

To determine the `zone_id`.

```yml
- name: fetch zone info
  markuman.hetzner_dns.zone_info:
    name: zone_name
```

# Dynamic Inventory

`hetzner_dns` collection comes with a dns based dynamic inventory.   
The file ending must be `.hetzner_dns.yml` or `.hetzner_dns.yaml`.

```yml
plugin: markuman.hetzner_dns.hetzner_dns
zone_name: osuv.de
```

Can be used with 

`ansible-inventory -i osuv.hetzner_dns.yml --list` or  
`ansible-playbook -i osuv.hetzner_dns.yml some_tasks.yml`

Supported filters are

```yml
plugin: markuman.hetzner_dns.inventory
zone_name: osuv.de
filters:
  type:
    - A
    - CNAME
```

The default type filter is set to `['A', 'AAAA', 'CNAME']`

### SCM

| **host** | **category** |
| --- | --- |
| https://git.osuv.de/m/hetzner_dns | origin |
| https://gitlab.com/markuman/hetzner_dns | pull mirror, issues, MR |
| https://github.com/markuman/hetzner_dns | push mirror, issues, PR |

#### License

GNU General Public License v3.0+  

_Basically I don't care about the license in this case. So I've decided to choose the same license as [official Ansible Hetzner Cloud Collection](https://github.com/ansible-collections/hetzner.hcloud)._


CHANGELOG.md
# 1.6.0

* Add dynamic inventory for hetzner dns

```yml
plugin: markuman.hetzner_dns.inventory
zone_name: osuv.de
filters:
  type:
    - A
    - CNAME
```

`ansible-inventory -i osuv.hetzner_dns.yml --list`

# 1.5.0

* fetch only requested `zone_name`
* improve integrationtests (_easier to run against custom domains_)

# 1.4.3

* fix bug. `purge: yes` removes not all existing records

# 1.4.2

* fix bug. even with `purge: yes`, it was always handled as `purge: no` and was always appending values.  
  *  https://git.osuv.de/m/hetzner_dns/commit/c57e5fd8b9a2a9ad8a9a3769a1f98e9664e0e685

# 1.4.1

* minor change (_documentation update_).

# 1.4.0

* add support to add/delete muliple DNS records for one Name
  * new paramter `purge` with alias parameter `replace`, `overwrite` and `solo` to be compatible with other ansible dns modules.


# 1.3.0

* Grafeful error handling
* full `--diff` support

---

# 1.2.0

* keep past record informations (_[Better Check Mode and Diff Support #1](https://github.com/markuman/hetzner_dns/issues/1)_)
* `check_mode` support also for `_info` modules (_[Better Check Mode and Diff Support #1](https://github.com/markuman/hetzner_dns/issues/1)_)
* Add GPL3+ License

---

# 1.1.0

Set default TTL to 300 to detect dns changes correctly. When a record was set manually without setting a TTL, the API does not response a TTL value.

plugins/module_utils/auth.py
import os
import requests
import json
from ansible.errors import AnsibleError

class HetznerAPIHandler:
    def __init__(self, kwargs):
        self.TOKEN = kwargs.get('api_token') or os.environ.get('HETZNER_DNS_TOKEN')
        if self.TOKEN is None:
            raise AnsibleError('Unable to continue. No Hetzner DNS API Token is given.')

    def get_zone_info(self):
        try:
            r = requests.get(
                url="https://dns.hetzner.com/api/v1/zones",
                headers={
                    "Auth-API-Token": self.TOKEN,
                },
            )

            if r.status_code == 200:
                return r
            else:
                raise AnsibleError('Unknown Error')
        except requests.exceptions.RequestException:
            raise AnsibleError('HTTP Request failed')

    def get_record_info(self, zone_id):
        try:
            r = requests.get(
                url="https://dns.hetzner.com/api/v1/records",
                params={
                    "zone_id": zone_id,
                },
                headers={
                    "Auth-API-Token": self.TOKEN,
                },
            )
            if r.status_code == 200:
                return r
            else:
                raise AnsibleError('Unknown Error')
        except requests.exceptions.RequestException:
            raise AnsibleError('HTTP Request failed')







plugins/modules/zone_info.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Part of ansible markuman.hetzner_dns collection

DOCUMENTATION = '''
module: markuman.hetzner_dns.zone_info
'''

EXAMPLES = '''
    - name: fetch zone info
      markuman.hetzner_dns.zone_info:
        name: zone_name
'''

from ansible.module_utils.basic import *
from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import HetznerAPIHandler
from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import ZoneInfo


def main():
    argument_spec = dict(
        name = dict(required=True, type='str'),
        api_token = dict(required=False, type='str', no_log=True, aliases=['access_token'])
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        supports_check_mode=True
    )

    dns = HetznerAPIHandler(module.params, module.fail_json)

    name = module.params.get("name")

    zones = dns.get_zone_info(name)
    zone_id, zone_info = ZoneInfo(zones, name)

    module.exit_json(changed = False, zone_id=zone_id, zone_info=zone_info)


if __name__ == '__main__':
    main()


plugins/inventory/inventory.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Part of ansible markuman.hetzner_dns collection


from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import HetznerAPIHandler
from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import ZoneInfo
from ansible.plugins.inventory import BaseInventoryPlugin
import os

DOCUMENTATION = '''
    name: hetzner_dns
    plugin_type: inventory

    options:
        zone_name:
            description: dns zone name.
            required: True
            type: string
        api_token:
            description: api token. if not set, it is read from env HETZNER_DNS_TOKEN
            required: False
            type: string
        filters:
            description:
                - A dictionary of filter value pairs.
            type: dict
            default: {}
            required: False
'''

class InventoryModule(BaseInventoryPlugin):
    NAME = 'markuman.hetzner_dns.inventory'

    def __init__(self):
        super(InventoryModule, self).__init__()
        # credentials
        self.api_token = self.get_option('api_token') or os.environ.get('HETZNER_DNS_TOKEN')

    def verify_file(self, path):
        valid = False
        if super(InventoryModule, self).verify_file(path):
            # base class verifies that file exists and is readable by current user
            if path.endswith(('hetzner_dns.yaml', 'hetzner_dns.yml')):
                valid = True
        return valid

    def parse(self, inventory, loader, path, cache=False):

        # call base method to ensure properties are available for use with other helper methods
        super(InventoryModule, self).parse(inventory, loader, path, cache)

        config = self._read_config_data(path)

        params = {'api_token': self.api_token}
        dns = HetznerAPIHandler(params)
        zone_name = self.get_option('zone_name')
        zones = dns.get_zone_info(zone_name)
        zone_id = ZoneInfo(zones, zone_name)

        records = dns.get_record_info(zone_id).json()['records']

        filters = self.get_option('filters')
        filter_types = filters.get('type') or ['A', 'AAAA', 'CNAME']

        #parse data and create inventory objects:
        for item in records:
            if item.get('type') in filter_types:
                name = item.get('name') + '.' + zone_name
                self.inventory.add_host(name)
                self.inventory.set_variable(name, 'ansible_host', item.get('value'))



plugins/module_utils/helper.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Part of ansible markuman.hetzner_dns collection

import os
import requests
import json
from ansible.errors import AnsibleError

error_codes = {
    200: "Successful response.",
    401: "Unauthorized.",
    403: "Forbidden.",
    404: "Not found.",
    406: "Not acceptable.",
    409: "Conflict.",
    422: "Unprocessable entity."
}
UE = "Unkown Error."

def ZoneInfo(zones, name):
    zone_id = None
    zone_info = {}
    for item in zones.json()['zones']:
        if item.get('name') == name:
            zone_id = item.get('id')
            zone_info = item

    return zone_id, zone_info

class HetznerAPIHandler:
    def __init__(self, params, fail_json=None):
        self.fail_json = fail_json
        self.TOKEN = params.get('api_token') or os.environ.get('HETZNER_DNS_TOKEN')
        if self.TOKEN is None:
            self.fail_json(msg='Unable to continue. No Hetzner DNS API Token is given.')

    def get_zone_info(self, zone_name):
        try:
            r = requests.get(
                url="https://dns.hetzner.com/api/v1/zones",
                headers={
                    "Auth-API-Token": self.TOKEN,
                },
                params={
                    "name": zone_name,
                },
            )

            if r.status_code == 200:
                return r
            else:
                self.fail_json(msg='zone_info: {msg}'.format(msg=error_codes.get(r.status_code, UE)))
        except requests.exceptions.RequestException:
            self.fail_json(msg='HTTP Request failed')

    def get_record_info(self, zone_id):
        try:
            r = requests.get(
                url="https://dns.hetzner.com/api/v1/records",
                params={
                    "zone_id": zone_id,
                },
                headers={
                    "Auth-API-Token": self.TOKEN,
                },
            )
            if r.status_code == 200:
                return r
            else:
                self.fail_json(msg='record_info: {msg}'.format(msg=error_codes.get(r.status_code, UE)))
        except requests.exceptions.RequestException:
            self.fail_json(msg='HTTP Request failed')


    def create_record(self, record):
        try:
            r = requests.post(
                url="https://dns.hetzner.com/api/v1/records",
                headers={
                    "Content-Type": "application/json",
                    "Auth-API-Token": self.TOKEN,
                },
                data=json.dumps(record)
            )
            if r.status_code == 200:
                return r
            else:
                raise AnsibleError('create record: {msg}'.format(msg=error_codes.get(r.status_code, UE)))
        except requests.exceptions.RequestException:
            raise AnsibleError('HTTP Request failed')


    def update_record(self, record, record_id):
        try:
            r = requests.put(
                url="https://dns.hetzner.com/api/v1/records/{RecordID}".format(RecordID=record_id),
                headers={
                    "Content-Type": "application/json",
                    "Auth-API-Token": self.TOKEN,
                },
                data=json.dumps(record)
            )

            if r.status_code == 200:
                return r
            else:
                self.fail_json(msg='update record: {msg}'.format(msg=error_codes.get(r.status_code, UE)))
        except requests.exceptions.RequestException:
            self.fail_json(msg='HTTP Request failed')

    def delete_record(self, record_id):
        try:
            r = requests.delete(
                url="https://dns.hetzner.com/api/v1/records/{RecordID}".format(RecordID=record_id),
                headers={
                    "Auth-API-Token": self.TOKEN,
                },
            )
            if r.status_code == 200:
                return r
            else:
                self.fail_json(msg='delete record: {msg}'.format(msg=error_codes.get(r.status_code, UE)))
        except requests.exceptions.RequestException:
            self.fail_json(msg='HTTP Request failed')












plugins/modules/record_info.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Part of ansible markuman.hetzner_dns collection

DOCUMENTATION = '''
module: markuman.hetzner_dns.zone_info
'''

EXAMPLES = '''
    - name: fetch zone info
      markuman.hetzner_dns.record_info:
        filter:
          - name: fritzbox
            value: 192.168.178.1
            type: A
          - name: fritzbox
            type: AAAA
        zone_name: osuv.de

'''

from ansible.module_utils.basic import *
from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import HetznerAPIHandler
from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import ZoneInfo


def main():
    argument_spec = dict(
        filter = dict(required=False, type='list'),
        zone_id = dict(required=False, type='str'),
        zone_name = dict(required=False, type='str'),
        api_token = dict(required=False, type='str', no_log=True, aliases=['access_token'])
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        mutually_exclusive=[['zone_id', 'zone_name']],
        required_one_of=[['zone_id', 'zone_name']],
        supports_check_mode=True
    )

    dns = HetznerAPIHandler(module.params, module.fail_json)

    filters = module.params.get("filter")
    zone_id = module.params.get("zone_id")
    zone_name = module.params.get("zone_name")

    if zone_id is None:
        zones = dns.get_zone_info(zone_name)
        zone_id = ZoneInfo(zones, zone_name)

    records = dns.get_record_info(zone_id)

    if filters is None:
        retval = records.json()['records']
    else:
        retval = list()
        for record in records.json()['records']:
            for filter in filters:
                if all(item in record.items() for item in filter.items()):
                    retval.append(record)



    module.exit_json(changed = False, record_info=retval)


if __name__ == '__main__':
    main()


plugins/inventory/hetzner_dns.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Part of ansible markuman.hetzner_dns collection


from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import HetznerAPIHandler
from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import ZoneInfo
from ansible.plugins.inventory import BaseInventoryPlugin
import os

DOCUMENTATION = '''
    name: hetzner_dns
    plugin_type: inventory

    options:
        zone_name:
            description: dns zone name.
            required: True
            type: string
        api_token:
            description: api token. if not set, it is read from env HETZNER_DNS_TOKEN
            required: False
            type: string
        filters:
            description:
                - A dictionary of filter value pairs.
            type: dict
            default: {}
            required: False
'''

class InventoryModule(BaseInventoryPlugin):
    NAME = 'markuman.hetzner_dns.hetzner_dns'

    def __init__(self):
        super(InventoryModule, self).__init__()
        # credentials
        self.api_token = self.get_option('api_token') or os.environ.get('HETZNER_DNS_TOKEN')

    def verify_file(self, path):
        valid = False
        if super(InventoryModule, self).verify_file(path):
            # base class verifies that file exists and is readable by current user
            if path.endswith(('hetzner_dns.yaml', 'hetzner_dns.yml')):
                valid = True
        return valid

    def parse(self, inventory, loader, path, cache=False):

        # call base method to ensure properties are available for use with other helper methods
        super(InventoryModule, self).parse(inventory, loader, path, cache)

        config = self._read_config_data(path)

        params = {'api_token': self.api_token}
        dns = HetznerAPIHandler(params)
        zone_name = self.get_option('zone_name')
        zones = dns.get_zone_info(zone_name)
        zone_id = ZoneInfo(zones, zone_name)

        records = dns.get_record_info(zone_id).json()['records']

        filters = self.get_option('filters')
        filter_types = filters.get('type') or ['A', 'AAAA', 'CNAME']

        #parse data and create inventory objects:
        for item in records:
            if item.get('type') in filter_types:
                name = item.get('name') + '.' + zone_name
                self.inventory.add_host(name)
                self.inventory.set_variable(name, 'ansible_host', item.get('value'))



plugins/modules/record.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
# Part of ansible markuman.hetzner_dns collection

DOCUMENTATION = '''
module: markuman.hetzner_dns.zone_info
'''

EXAMPLES = '''
    - name: add record
      markuman.hetzner_dns.record:
        zone_name: osuv.de
        name: hetzner_dns_ansible_collection
        value: osuv.de.
        type: CNAME
        ttl: 300
      register: RECORD

    - name: add record no change
      markuman.hetzner_dns.record:
        zone_name: osuv.de
        name: hetzner_dns_ansible_collection
        value: osuv.de.
        type: CNAME
        ttl: 300
      register: RECORD

    - name: add record change
      markuman.hetzner_dns.record:
        zone_name: osuv.de
        name: hetzner_dns_ansible_collection
        value: osuv.de.
        type: CNAME
        ttl: 60
      register: RECORD

    - name: del record
      markuman.hetzner_dns.record:
        zone_name: osuv.de
        name: hetzner_dns_ansible_collection
        type: CNAME
        state: absent
      register: RECORD
'''

from ansible.module_utils.basic import *
from ansible.errors import AnsibleError
from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import HetznerAPIHandler
from ansible_collections.markuman.hetzner_dns.plugins.module_utils.helper import ZoneInfo, error_codes, UE
import yaml

def main():
    argument_spec = dict(
        zone_id = dict(required=False, type='str'),
        zone_name = dict(required=False, type='str'),
        api_token = dict(required=False, type='str', no_log=True, aliases=['access_token']),
        name = dict(required=True, type='str'),
        value = dict(type='str'),
        ttl = dict(default=300, type='int'),
        type = dict(required=True, type='str', choices=["A","AAAA","NS","MX","CNAME","RP","TXT","SOA","HINFO","SRV","DANE","TLSA","DS","CAA"]),
        state = dict(type='str', default='present', choices=['present', 'absent']),
        purge = dict(required=False, type='bool', default=True, aliases=['replace', 'overwrite', 'solo'])
    )

    module = AnsibleModule(
        argument_spec=argument_spec,
        required_if=[['state', 'present', ['value']]],
        required_one_of=[
            ['zone_id', 'zone_name']
        ],
        supports_check_mode=True
    )

    dns = HetznerAPIHandler(module.params, module.fail_json)

    zone_id = module.params.get("zone_id")
    zone_name = module.params.get("zone_name")
    state = module.params.get("state")
    purge = module.params.get("purge")

    if zone_id is None:
        zones = dns.get_zone_info(zone_name)
        zone_id, zone_info = ZoneInfo(zones, zone_name)
        if zone_id is None:
          module.fail_json(msg='zone or zone_id: {msg}'.format(msg=error_codes.get(404, UE)))

    future_record = {
        'name': module.params.get("name"),
        'value': module.params.get("value"),
        'type': module.params.get("type"),
        'ttl': int(module.params.get("ttl")),
        'zone_id': zone_id
    }

    find_record = {
        'name': future_record.get('name'),
        'type': future_record.get('type')
    }

    if module.params.get("value") and not purge:
        find_record['value'] = module.params.get("value")

    records = dns.get_record_info(zone_id)

    record_changed = False
    record_exists = False
    change = False
    past_record = []
    record_ids = []
    for record in records.json()['records']:

        if not record.get('ttl'):
          record['ttl'] = 300

        if all(item in record.items() for item in find_record.items()):
            record_exists = True
            record_id = record.get('id')
            record_ids.append(record.get('id'))
            past_record.append(record)

            if not all(item in record.items() for item in future_record.items()):
                record_changed = True
            else:
                this_record = { 'record': record }


    if record_changed and not purge:
        record_exists = False
        record_changed = False

    if state == 'present':
        if purge and len(record_ids) > 1:
            for id in record_ids:
                r = dns.delete_record(id)

        if not record_exists or (purge and len(record_ids) > 1):
            record_id = None
            this_record = { 'record': future_record }
            change = True

            if not module.check_mode:
                r = dns.create_record(future_record)
                record_id = r.json()['record']['id']
                this_record = r.json()

        elif record_changed:
            change = True
            this_record = { 'record': future_record }
            if not module.check_mode:
              r = dns.update_record(future_record, record_id)
              this_record = r.json()

    if state == 'absent':
        if record_exists:
            change = True
            if not module.check_mode:
                if not module.params.get("value"):
                    # when value was not defined, delete all records of this type
                    for id in record_ids:
                        r = dns.delete_record(id)
                else:
                    r = dns.delete_record(record_id)

        else:
            change = False
        this_record = { 'record': {} }
        record_id = None

    for idx in range(len(past_record)):
        past_record[idx].pop('id', None)
        past_record[idx].pop('created', None)
        past_record[idx].pop('modified', None)
        past_record[idx].pop('zone_id', None)

    after_record = [this_record.get('record')]
    if isinstance(after_record[0], dict):
        after_record[0].pop('id', None)
        after_record[0].pop('created', None)
        after_record[0].pop('modified', None)
        after_record[0].pop('zone_id', None)

    diff = dict(
        before=yaml.safe_dump(past_record),
        after=yaml.safe_dump(after_record)
    )

    module.exit_json(changed = change, record_id=record_id, record_info=this_record, past_record=past_record, diff=diff)


if __name__ == '__main__':
    main()
